### PR TITLE
Stream-friendly smart playlists

### PR DESCRIPTION
**Pull request by catap**
## Description

This commit introduced a way to generate a stream-frienldy playlists.

This is that I've missed to make me very happy at #3847

## To Do

- [x] Documentation. (If you've add a new command-line flag, for example, find the appropriate page under `docs/` to describe it.)
- [ ] Changelog. (Add an entry to `docs/changelog.rst` near the top of the document.)
- [ ] Tests. (Encouraged but not strictly required.)
- [x] ~figured out that to do with complex unicode strings.~

## ~Unicode-related issue~

~Right now current code encodes at least at python `3.8.7` a string `Ayọ` as `Ayo%CC%A3` when it should be `Ay%E1%BB%8D`.~

~Until it is fixed I ask to don't merge it. @sampsyo if you have any idea how to fix it I'll be appreciated.~

### CODE REVIEW

**Comment by catap** [comment]
Status update here. I guess I discovered a wired bug in macOS:

It reads from FS as o plus modificator, as result I do have:
```
>>> from urllib.request import pathname2url
>>> pathname2url(u'Ayọ')
'Ayo%CC%A3'
>>> pathname2url(u'Ayọ')
'Ay%E1%BB%8D'
>>> 
```

**Comment by catap** [comment]
Seems like using `asciify_paths` is a good way to do not care about unicode here.

File: beetsplug/smartplaylist.py [hunk 1]
```diff
@@ -39,8 +44,12 @@ def __init__(self):
             'auto': True,
             'playlists': [],
             'forward_slash': False,
+            'started_with': u'',
+            'urlencoded': False,
```
Comment (sampsyo) [inline beetsplug/smartplaylist.py:48]: Can we just call this `urlencode`?

File: beetsplug/smartplaylist.py [hunk 2]
```diff
@@ -39,8 +44,12 @@ def __init__(self):
             'auto': True,
             'playlists': [],
             'forward_slash': False,
+            'started_with': u'',
```
Comment (sampsyo) [inline beetsplug/smartplaylist.py:47]: Can we call this `prefix`?

File: beetsplug/smartplaylist.py [hunk 3]
```diff
@@ -201,6 +210,7 @@ def update_playlists(self, lib):
                 if item_path not in m3us[m3u_name]:
                     m3us[m3u_name].append(item_path)
 
+        started_with = bytestring_path(self.config['started_with'].get())
```
Comment (sampsyo) [inline beetsplug/smartplaylist.py:213]: I don't think `bytestring_path` is the right thing here because this isn't necessarily a path. I think we should encode it with whatever the text encoding is for the playlist file?

File: beetsplug/smartplaylist.py [hunk 4]
```diff
@@ -210,6 +220,8 @@ def update_playlists(self, lib):
                 for path in m3us[m3u]:
                     if self.config['forward_slash'].get():
                         path = path_as_posix(path)
-                    f.write(path + b'\n')
+                    if self.config['urlencoded'].get():
```
Comment (sampsyo) [inline beetsplug/smartplaylist.py:223]:
```suggestion
                    if self.config['urlencoded']:
```

File: beetsplug/smartplaylist.py [hunk 5]
```diff
@@ -210,6 +220,8 @@ def update_playlists(self, lib):
...
+                        path = bytestring_path(pathname2url(path))
```
Comment (sampsyo) [inline beetsplug/smartplaylist.py:224]: Again, I'm not sure `bytestring_path` is the right thing here…

File: docs/plugins/smartplaylist.rst [hunk 6]
```diff
@@ -101,3 +121,6 @@ other configuration options are:
   If you intend to use this plugin to generate playlists for MPD on
   Windows, set this to yes.
   Default: Use system separator.
+- **started_with**: Started each path to track with specified string, for
+  example path to server where you host your music. Default empty string.
```
Comment (sampsyo) [inline docs/plugins/smartplaylist.rst:125]: > Prepend this string to every path in the playlist file. For example, you could use the URL for a server where the music is stored. Default: empty string.

File: docs/plugins/smartplaylist.rst [hunk 7]
```diff
@@ -101,3 +121,6 @@ other configuration options are:
...
+- **urlencoded**: URL encode all paths. Default ``no``.
```
Comment (sampsyo) [inline docs/plugins/smartplaylist.rst:126]:
```suggestion
- **urlencoded**: URL-encode all paths. Default: ``no``.
```

File: docs/plugins/smartplaylist.rst [hunk 8]
```diff
@@ -30,6 +30,26 @@ use spaces in the query, be sure to quote them (e.g., ``artist:"The Beatles"``).
 If you have existing files with the same names, you should back them up---they
 will be overwritten when the plugin runs.
 
+You also can generate a playlists to play music on your NAS or any server. Let
+assume that you host all your music at ``https://nas.domain.com/music``, that is
+mounted to your machine when you run beet as ``/mnt/music`` and you should login
+to it as user ``user`` with password ``secret``::
```
Comment (sampsyo) [inline docs/plugins/smartplaylist.rst:36]: Could we put this in a separate section, because it's separate from the basic instructions? Also, I'm not sure we need an example config with all the other settings, such as `playlists:`. Can we keep the example compact (if an example is necessary at all)?

**Review by sampsyo — COMMENTED** [review]
Seems interesting; thanks for getting this started. Could you please add a changelog entry?

File: beetsplug/smartplaylist.py (hunk 4 above)
Comment (catap) [reply beetsplug/smartplaylist.py:223]: Hm, shall I change `forward_slash` also?

File: beetsplug/smartplaylist.py (hunk 5 above)
Comment (catap) [reply beetsplug/smartplaylist.py:224]: @sampsyo without it CI fails ;)

Let me show it as separated commit

File: docs/plugins/smartplaylist.rst (hunk 8 above)
Comment (catap) [reply docs/plugins/smartplaylist.rst:36]: Let me remove this example because it is tricky one.

**Comment by catap** [comment]
@sampsyo ping here

File: beetsplug/smartplaylist.py [hunk 9]
```diff
@@ -210,6 +219,8 @@ def update_playlists(self, lib):
                 for path in m3us[m3u]:
                     if self.config['forward_slash'].get():
                         path = path_as_posix(path)
-                    f.write(path + b'\n')
+                    if self.config['urlencode'].get():
```
Comment (sampsyo) [inline beetsplug/smartplaylist.py:222]:
```suggestion
                    if self.config['urlencode']:
```

File: beetsplug/smartplaylist.py [hunk 10]
```diff
@@ -39,8 +44,12 @@ def __init__(self):
             'auto': True,
             'playlists': [],
             'forward_slash': False,
+            'prefix': u'',
+            'urlencode': False,
         })
 
+        # may contains user and password
+        self.config['prefix'].redact = True
```
Comment (sampsyo) [inline beetsplug/smartplaylist.py:52]:
```suggestion
        self.config['prefix'].redact = True  # May contain username/password.

```

Preferring a blank line afterward because this is unrelated to the next two lines.

**Review by sampsyo — CHANGES_REQUESTED** [review]
Looking pretty good! Could you please add a changelog entry?

**Comment by catap** [comment]
@sampsyo ;)

File: beetsplug/smartplaylist.py [hunk 11]
```diff
@@ -201,6 +209,7 @@ def update_playlists(self, lib):
                 if item_path not in m3us[m3u_name]:
                     m3us[m3u_name].append(item_path)
 
+        started_with = bytestring_path(self.config['prefix'].get())
```
Comment (sampsyo) [inline beetsplug/smartplaylist.py:212]: `get` should be `as_str` here to enforce a type on the configuration. And can we call this variable `prefix`?

**Review by sampsyo — CHANGES_REQUESTED** [review]
Oops; noticed one last small thing (or two)! Sorry for letting this slip by last time. This should be good to go afterward.

File: beetsplug/smartplaylist.py [hunk 12]
```diff
@@ -28,6 +28,11 @@
 import os
 import six
 
+try:
+    from urllib.request import pathname2url
+except ImportError:
+    from urllib import pathname2url
+
```
Comment (justinmayer) [inline beetsplug/smartplaylist.py:35]: I suggest adding a comment above to indicate which Python version(s) this compatibility block is intended to support. That will make it easier to find and remove this stanza when Beets drops support for said Python versions(s) in the future.

File: docs/changelog.rst [hunk 13]
```diff
@@ -2066,6 +2066,9 @@ And many little fixes and improvements:
 * When there's a parse error in a query (for example, when you type a
   malformed date in a :ref:`date query <datequery>`), beets now stops with an
   error instead of silently ignoring the query component.
+* :doc:`/plugins/smartplaylist`: Stream-friendly smart playlists.
+  The ``splupdate`` command can now also add some prefix to every path
+  in the playlist file which can be URL-encode.
```
Comment (justinmayer) [inline docs/changelog.rst:2071]: Suggested re-phrasing:
```
The ``splupdate`` command can now also add a URL-encodable prefix to every path
in the playlist file.
```

**Review by justinmayer — COMMENTED** [review]
Nice work, Kirill. I made a few minor suggestions. 💫

**Comment by catap** [comment]
I defently forgot about it. Will try to finish in few days.

**Comment by catap** [comment]
@sampsyo / @justinmayer a few days was a bit longer... well.. It is fixed! :)

**Review by sampsyo — APPROVED** [review]
Looks great! Thanks for seeing it though!!
